tests/test_transpilers_routing.py

```diff
@@ -0,0 +1,239 @@
+    circuit.add(gates.CZ(0, 1))
+    assert not respect_connectivity(star_connectivity(), circuit, verbose)
+
+
+@pytest.mark.parametrize("verbose", [True, False])
+def test_respect_connectivity_3q(verbose):
+    circuit = Circuit(5)
+    circuit.add(gates.TOFFOLI(0, 1, 2))
+    assert not respect_connectivity(star_connectivity(), circuit, verbose)
+
+
+def test_remap_circuit():
+    circuit = Circuit(3)
+    circuit.add(gates.CNOT(1, 0))
+    circuit.add(gates.CNOT(2, 0))
+    circuit.add(gates.CNOT(1, 2))
+    qubit_map = np.asarray([2, 1, 0])
+    ref_circuit = Circuit(3)
+    ref_circuit.add(gates.CNOT(1, 2))
+    ref_circuit.add(gates.CNOT(0, 2))
+    ref_circuit.add(gates.CNOT(1, 0))
+    new_circuit = remap_circuit(circuit, qubit_map)
+    for i, gate in enumerate(new_circuit.queue):
+        assert gate.qubits == ref_circuit.queue[i].qubits
+
+
+@pytest.mark.parametrize("split", [2.0, -1.0])
+def test_split_setter(split):
+    with pytest.raises(ValueError):
+        transpiler = ShortestPaths(connectivity=star_connectivity(), sampling_split=split)
+
+
+def test_insufficient_qubits():
+    circuit = generate_random_circuit(10, 20)
+    placer = Trivial()
+    initial_layout = placer(circuit)
+    transpiler = ShortestPaths(connectivity=star_connectivity())
+    with pytest.raises(ValueError):
+        transpiler(circuit, initial_layout)
+
+
+def test_incorrect_initial_layout():
+    placer = Trivial()
+    circuit = Circuit(4)
+    circuit.add(gates.CNOT(1, 0))
+    circuit.add(gates.CNOT(2, 0))
+    circuit.add(gates.CNOT(3, 0))
+    initial_layout = placer(circuit)
+    transpiler = ShortestPaths(connectivity=star_connectivity())
+    with pytest.raises(ValueError):
+        transpiler(circuit, initial_layout)
+
+
+@pytest.mark.parametrize("gates", [1, 10, 50])
+@pytest.mark.parametrize("qubits", [3, 5])
+def test_random_circuits_5q(gates, qubits):
+    placer = Trivial()
+    layout_circ = Circuit(5)
+    initial_layout = placer(layout_circ)
+    transpiler = ShortestPaths(connectivity=star_connectivity(), verbose=True)
+    circuit = generate_random_circuit(nqubits=qubits, ngates=gates)
+    transpiled_circuit, final_qubit_map = transpiler(circuit, initial_layout)
+    assert transpiler.added_swaps >= 0
+    assert transpiler.is_satisfied(transpiled_circuit)
+    assert respect_connectivity(star_connectivity(), transpiled_circuit)
+    assert check_placement(transpiled_circuit, final_qubit_map)
+
+
+def q21_connectivity():
+    Q = ["q" + str(i) for i in range(21)]
+    chip = nx.Graph()
+    chip.add_nodes_from(Q)
+    graph_list_h = [
+        (Q[0], Q[1]),
+        (Q[1], Q[2]),
+        (Q[3], Q[4]),
+        (Q[4], Q[5]),
+        (Q[5], Q[6]),
+        (Q[6], Q[7]),
+        (Q[8], Q[9]),
+        (Q[9], Q[10]),
+        (Q[10], Q[11]),
+        (Q[11], Q[12]),
+        (Q[13], Q[14]),
+        (Q[14], Q[15]),
+        (Q[15], Q[16]),
+        (Q[16], Q[17]),
+        (Q[18], Q[19]),
+        (Q[19], Q[20]),
+    ]
```

Comment:
```suggestion
    graph_list_h = [(Q[i], Q[i+1]) for i in range(20)]
```

Reply:
```suggestion
    graph_list_h = [(Q[i], Q[i+1]) for i in range(20) if i % 5 != 2]
```

Reply:
> if you look closely it is not the same thing...

This is a good reason for rewriting: with the full list, it is difficult to spot the features. When you spell out the criterion instead, it's fully explicit.